### ISSUE TITLE
Can we allow making everything on gpu/cuda for scheduler?

### ISSUE THREAD

**Issue by xiang9156** [huggingface/diffusers#9485]
**What API design would you like to have changed or added to the library? Why?**
In https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lcm.py
It looks like that attributes like `scheduler.alphas_cumprod` are tensors on cpu, but the scheduler.set_timesteps() allows setting `scheduler.timesteps` to gpu/cuda device. Isn't this causing device mismatch when indexing scheduler.alphas_cumprod with scheduler.timesteps? Below is the code snippet that the pipline is indexing a cpu tensor(alphas_cumprod) with a gpu tensor(timestep)
![image](https://github.com/user-attachments/assets/42b31655-0b4f-4623-9524-5d55bf7b7f5c)
I simply added following lines to print the timestep and self.alphas_cumprod
```
print("Printing scheduler.step() timestep")
print(type(timestep))
print(isinstance(timestep, torch.Tensor))
print(timestep.device)
print("Printing scheduler.step() self.alphas_cumprod")
print(type(self.alphas_cumprod))
print(isinstance(self.alphas_cumprod, torch.Tensor))
print(self.alphas_cumprod.device)
``` 
Output when running text-to-image:
```
Printing scheduler.step() timestep
<class 'torch.Tensor'>
True
cuda:0
Printing scheduler.step() self.alphas_cumprod
<class 'torch.Tensor'>
True
cpu
```

**What use case would this enable or better enable? Can you give us a code example?**
We are using LCMScheduler for video generations, it's generating frames repeatedly in a loop. for most of time, this step doesn't cause performance issue. But we did see intermittent high cpu usage and latency for `alpha_prod_t = self.alphas_cumprod[timestep]`. And from torch.profiler and tracing output, it. shows high latency for this specific step. We are wondering if this is the performance bottleneck.
![image](https://github.com/user-attachments/assets/e8cb0cfe-bed7-440e-875f-a133892e7f36)

**Comment by asomoza** [comment]
cc: @yiyixuxu

**Comment by yiyixuxu** [comment]
thanks for the issue! and interesting that this performance issue is not deterministic 
 cc @a-r-r-o-w here, I think cogvideox scheduler has a similar design and he is investigating for a similar issue

note that most of our popular schedulers (euler/dpm), step calculation only replied on sigmas (which we put on cpu and use an `int` to index, which seems to be the best option performance-wise, with and without `torch.compile` https://github.com/huggingface/diffusers/pull/6173)

we can not initiate everything on GPU because the user device is not known at the point of initiation 

```
# scheduler.__init__ is called at this step
pipe = DiffusionPipeline.from(...) 
# user can then enable offloading or move to a device
pipe.enable_model_cpu_offload() # or pipe.to("cuda")
# we wouldn't know the device until the pipeline is called 
out = pipe(...)
```

maybe we see if we can update the calculation to be based on sigma?

**Comment by xiang9156** [comment]
thanks! the official diffusers document loads LCMScheduler in this way
![image](https://github.com/user-attachments/assets/ea9f127b-c750-410f-869d-68a9f6e738e1)
can we have extra arg like device to make it initialize with passed in device?

just curious, if we keep everything on cpu for scheduler, wouldn't unet model implicitly convert timestep into tensor on gpu device during inference?

> thanks for the issue! and interesting that this performance issue is not deterministic cc @a-r-r-o-w here, I think cogvideox scheduler has a similar design and he is investigating for a similar issue
> 
> note that most of our popular schedulers (euler/dpm), step calculation only replied on sigmas (which we put on cpu and use an `int` to index, which seems to be the best option performance-wise, with and without `torch.compile` #6173)
> 
> we can not initiate everything on GPU because the user device is not known at the point of initiation
> 
> ```
> # scheduler.__init__ is called at this step
> pipe = DiffusionPipeline.from(...) 
> # user can then enable offloading or move to a device
> pipe.enable_model_cpu_offload() # or pipe.to("cuda")
> # we wouldn't know the device until the pipeline is called 
> out = pipe(...)
> ```
> 
> maybe we see if we can update the calculation to be based on sigma?

**Comment by a-r-r-o-w** [comment]
Hi @xiang9156. I shared my investigation in [this](https://github.com/huggingface/diffusers/pull/9475#issuecomment-2367852119) comment. Do your observations align with it? We're thinking about how to remove all cuda stream synchronization from the pipeline (which should speed things up quite a bit across all pipelines), so if you have any other insights from your own investigation(s), please feel free to chime in and we can jam on implementing the improvements together.

**Comment by xiang9156** [comment]
Hi @a-r-r-o-w , i saw your investigation, thanks for doing that!! I'll also try moving everythong onto cpu/gpu on my end to see the performance gain.

**Comment by xiang9156** [comment]
> Hi @xiang9156. I shared my investigation in [this](https://github.com/huggingface/diffusers/pull/9475#issuecomment-2367852119) comment. Do your observations align with it? We're thinking about how to remove all cuda stream synchronization from the pipeline (which should speed things up quite a bit across all pipelines), so if you have any other insights from your own investigation(s), please feel free to chime in and we can jam on implementing the improvements together.
> 
> Another area for improvement could be removing unnecessary calls to cudaMemCpy which seem like the next obvious bottleneck to tackle after fighting cudaStreamSync's.
@a-r-r-o-w 

i just tried very simple script on Google Colab to compare 4 cases:
- indexing a cpu tensor with gpu tensor scalar
- indexing a cpu tensor with cpu tensor scalar (Fastest)
- indexing a gpu tensor with cpu tensor scalar
- indexing a gpu tensor with gpu tensor scalar

It did show some useful information to me: indexing a gpu tensor with gpu tensor scalar is as slow as other 2 mismatched device cases, 
![image](https://github.com/user-attachments/assets/d1b03fda-2a22-4641-a214-957c3f96ed8b)

**Comment by xiang9156** [comment]
> > Hi @xiang9156. I shared my investigation in [this](https://github.com/huggingface/diffusers/pull/9475#issuecomment-2367852119) comment. Do your observations align with it? We're thinking about how to remove all cuda stream synchronization from the pipeline (which should speed things up quite a bit across all pipelines), so if you have any other insights from your own investigation(s), please feel free to chime in and we can jam on implementing the improvements together.
> > Another area for improvement could be removing unnecessary calls to cudaMemCpy which seem like the next obvious bottleneck to tackle after fighting cudaStreamSync's.
> > @a-r-r-o-w
> 
> i just tried very simple script on Google Colab to compare 4 cases:
> 
> * indexing a cpu tensor with gpu tensor scalar
> * indexing a cpu tensor with cpu tensor scalar (Fast)
> * indexing a gpu tensor with gpu tensor scalar
> * indexing a gpu tensor with cpu tensor scalar(Fast)
> 
> It did show some useful information to me: indexing a gpu tensor with gpu tensor scalar is super slow... i didn't get time to do torch profiling but i guess it might be caused by synchronization, i may do that later today. ![image](https://private-user-images.githubusercontent.com/14140862/370188010-47c0cf24-2926-4bab-a247-e30648993857.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjU1NzIsIm5iZiI6MTcyNzE2NTI3MiwicGF0aCI6Ii8xNDE0MDg2Mi8zNzAxODgwMTAtNDdjMGNmMjQtMjkyNi00YmFiLWEyNDctZTMwNjQ4OTkzODU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDA4MDc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0NjVkMWNiNWExOGI3MTM1OGU5ZTY1MDBiYjEzZGU4ZDJlNThkZTUwMmViMDkzY2U3YzRjZDU1ODkwZTFlYjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GkZbhUiOPgyQ0ObEkt5xCw-lnkfKKx0KbZ2GtxtOzI4)

i tried profiling our own scheduler after setting timesteps on cpu, the scheduler.step() latency dopped to us level(hundreds of us). Without setting timesteps on cpu, the scheduler.step() latency was ranging from 3ms to 100+ms(happened intermittently for video generation). i'm trying to find any side effect of setting timesteps on cpu

**Comment by xiang9156** [comment]
from tracing, i did see extra sync in controlnet forward and unet forward after i set timestep tensor to cpu
below is the extra controlnet forward tracing
![image](https://github.com/user-attachments/assets/a6d6a663-fb3b-4a39-9e0d-df43e368f0f7)
![image](https://github.com/user-attachments/assets/3a2ab464-2298-4f15-8913-456163a0bada)

**Comment by xiang9156** [comment]
> I think it's better to maintain two copies of timesteps in the scheduler for this (one on cuda and one on cpu). To the controlnet/unet/transformer, you can pass the cuda timestep but to the scheduler, you can pass the cpu timestep. This makes it so that there are no cuda synchronizations at all until completion of inference.
> 
> However, I'm only now noticing that even if we save time from cuda synchronization by using above method, we will have to sync at some point in time before/after decoding. This causes roughly the same inference time overall leading to almost no performance gains. I'll post a more detailed comment soon in the PR

Yes, i'm also setting two copies of timesteps. the sync did disappear.
